src/ngx_http_lua_upstream_module.c

```diff
@@ -349,6 +353,65 @@ ngx_http_lua_upstream_set_peer_down(lua_State * L)
 }
 
 
+static int
+ngx_http_lua_upstream_get_next_peer(lua_State * L)
+{
+    ngx_str_t                           host;
+    ngx_http_upstream_srv_conf_t       *us;
+    ngx_http_request_t                 *r;
+    ngx_http_request_t                  fake_r;
+    ngx_peer_connection_t              *peer;
+    int                                 rc;
+
+    if (lua_gettop(L) != 1) {
+        return luaL_error(L, "exactly one argument expected");
+    }
+
+    host.data = (u_char *) luaL_checklstring(L, 1, &host.len);
+
+    us = ngx_http_lua_upstream_find_upstream(L, &host);
+    if (us == NULL) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "upstream not found");
+        return 2;
+    }
+
+    r = ngx_http_lua_get_request(L);
+
+    /* we create a fake request object, which we will use to fetch the upstream
+     * peer, because we don't want to pollute the current request object */
+    ngx_memzero(&fake_r, sizeof(ngx_http_request_t));
+    fake_r.pool = r->pool;
+    fake_r.connection = r->connection;
+    rc = ngx_http_upstream_create(&fake_r);   /* setup fake_r.upstream */
+    if (rc != NGX_OK) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "could not determine next peer");
+        return 2;
+    }
+
+    /* calls the configured peer chooser module (ip_hash/hash/round-robin) init
+     * function, which sets up fake_r.upstream->peer */
+    rc = us->peer.init(&fake_r, us);
+    if (rc != NGX_OK) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "could not determine next peer");
+        return 2;
+    }
+
+    peer = &fake_r.upstream->peer;
+    rc = peer->get(peer, peer->data);
+    if (rc != NGX_OK) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "could not determine next peer");
+        return 2;
+    }
+
```

Comment:
Maybe we should call `peer->free` here to ensure any resources allocated in `peer->get` are freed properly?